rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +137,232 @@ public void close(ISourceLocation loc) {
         }
     }
 
+    /**
+     * This class keeps together diagnostics messages and the versioned syntax
+     * tree for which the diagnostics were computed. This is needed to ensure
+     * that builder diagnostics are calculated for the same syntax as
+     * supposedly-corresponding analyzer diagnostics. The diagnostics can come
+     * from any source (e.g., parser, analyzer, builder).
+     */
+    private static class VersionedDiagnostics {
+        public final Versioned<ITree> tree;
+        public final List<Diagnostic> messages;
+
+        public VersionedDiagnostics(Versioned<ITree> tree, List<Diagnostic> messages) {
+            this.tree = tree;
+            this.messages = messages;
+        }
+
+        private static final VersionedDiagnostics EMPTY = new VersionedDiagnostics(
+            new Versioned<>(-1, IRascalValueFactory.getInstance().character(0)),
+            new ArrayList<>());
```

Comment:
NP: `Collections.emptyList()` is nice for these cases, as it also prevents anyone adding stuff to it.